### PR TITLE
JAMES-3362 Jmap VacationResponse/get implementation

### PR DESCRIPTION
**Pull request by Arsnael**

### CODE REVIEW

**Comment by Arsnael** [comment]
I think the PR gets stable enough, so I switched it to review mode. I still let the label `WIP` as I still miss properties filtering, but the rest should be here

File: server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/json/VacationResponseSerializationTest.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+class VacationResponseSerializationTest extends AnyWordSpec with Matchers {
+  "Serialize VacationResponse" should {
+    "succeed" in {
+      val expectedJson: String =
+        """{
+          | "id":"singleton",
+          | "isEnabled":true,
+          | "fromDate":"2016-10-09T08:07:06+07:00[Asia/Vientiane]",
+          | "toDate":"2017-10-09T08:07:06+07:00[Asia/Vientiane]",
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/json/VacationResponseSerializationTest.scala:63]: This needs to be expressed with an UTCDate:

https://jmap.io/spec-core.html#the-date-and-utcdate-data-types

```
Where UTCDate is given as a type, it means a Date where the time-offset component MUST be Z (i.e., it must be in UTC time). For example, "2014-10-30T06:12:00Z".
```

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/Serializer.scala [hunk 2]
```diff
@@ -407,7 +417,10 @@ class Serializer @Inject() (mailboxIdFactory: MailboxId.Factory) {
 
   def serialize(errors: JsError): JsValue = Json.toJson(errors)
 
-  def serialize(vacationResponse: VacationResponse): JsValue = Json.toJson(vacationResponse)
+  def serialize(vacationResponse: VacationResponse)(implicit vacationResponseWrites: Writes[VacationResponse]): JsValue =
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/Serializer.scala:420]: `(implicit vacationResponseWrites: Writes[VacationResponse])` is likely not needed?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/VacationResponseGetMethod.scala [hunk 3]
```diff
@@ -47,20 +47,29 @@ class VacationResponseGetMethod @Inject() (serializer: Serializer,
...
+                list = List(vacationResult.vacationResponse),
+                notFound = vacationResult.notFound))
+              .map(vacationResponseGetResponse => Invocation(
+                methodName = methodName,
+                arguments = Arguments(serializer.serialize(vacationResponseGetResponse).as[JsObject]),
+                methodCallId = invocation.methodCallId)))))
+
   }
 
+  private def validate(capabilities: Set[CapabilityIdentifier], invocation: Invocation): SMono[Invocation] =
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/VacationResponseGetMethod.scala:78]: Maybe we can use Either for this validation step?

**Review by chibenwa — COMMENTED** [review]
- Do we have property filtering implemented? (maybe in a follow up pull request)

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/VacationResponseGetMethod.scala (hunk 3 above)
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/VacationResponseGetMethod.scala:78]: Will try

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/Serializer.scala (hunk 2 above)
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/json/Serializer.scala:420]: The result of a bad rebase, thanks :)

**Comment by Arsnael** [comment]
This work is big enough already and self-sufficient. Will handle the properties filtering in an other following PR

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/VacationResponseGetMethodContract.scala [hunk 4]
```diff
@@ -145,8 +145,8 @@ trait VacationResponseGetMethodContract {
          |        {
          |          "id":"singleton",
          |          "isEnabled": true,
-         |          "fromDate": "2014-09-30T14:10:00+02:00",
-         |          "toDate": "2016-04-15T11:56:32.224+07:00[Asia/Vientiane]",
+         |          "fromDate": "2014-09-30T12:10:00Z[UTC]",
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/VacationResponseGetMethodContract.scala:148]: Should be 2014-09-30T12:10:00Z (without [UTC]) as far as I understood.

Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/VacationResponseGetMethodContract.scala:148]: Seems so... Might need to use a custom DateTimeFormatter then I guess in the Serializer, will check it out

**Comment by Arsnael** [comment]
```
[8b4ada386ea9ffa67d5942f841588d389fa39ada] - should succeed *** FAILED ***
[8b4ada386ea9ffa67d5942f841588d389fa39ada]   net.javacrumbs.jsonunit.core.internal.JsonAssertError: JSON documents are different:
[8b4ada386ea9ffa67d5942f841588d389fa39ada] Different value found in node "fromDate", expected: <"2016-10-09T01:07:06Z[UTC]"> but was: <"2016-10-09T01:07:06Z">.
[8b4ada386ea9ffa67d5942f841588d389fa39ada] Different value found in node "toDate", expected: <"2017-10-09T01:07:06Z[UTC]"> but was: <"2017-10-09T01:07:06Z">.
```
ok forgot something obviously :)

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/VacationResponseGetMethodContract.scala [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+         |          "id":"singleton",
+         |          "isEnabled": false
+         |        }
+         |      ],
+         |      "notFound": ["random1", "random2"]
+         |    },
+         |    "c1"]]
+         |}""".stripMargin)
+  }
+}
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/VacationResponseGetMethodContract.scala:330]: Can you add a test to see what happes when I do request an empty id : 

```
    val response = `given`
      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
      .body(s"""{
               |  "using": [
               |    "urn:ietf:params:jmap:core",
               |    "urn:ietf:params:jmap:mail",
               |    "urn:ietf:params:jmap:vacationresponse"],
               |  "methodCalls": [[
               |    "VacationResponse/get",
               |    {
               |      "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
               |      "ids": [""]
               |    },
               |    "c1"]]
               |}""".stripMargin)
    .when
      .post
    .`then`
      .statusCode(SC_OK)
      .contentType(JSON)
      .extract
      .body
      .asString
```

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/VacationResponseGetMethod.scala [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  private def validate(capabilities: Set[CapabilityIdentifier]): Either[MissingCapabilityException, Unit] =
+    if (!capabilities.contains(CapabilityIdentifier.JMAP_VACATION_RESPONSE)) {
+      Left(MissingCapabilityException(CapabilityIdentifier.JMAP_VACATION_RESPONSE))
+    } else {
+      Right()
+    }
+
+  private def asVacationResponseGetRequest(arguments: Arguments): SMono[VacationResponseGetRequest] =
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/VacationResponseGetMethod.scala:86]: If the return type is an either, we would be able to write:

```
validate(capabilities)
    .flatMap(_ => asVacationResponseGetRequest(invocation.arguments))
    .fold(...)
```

I think we should rely on Either when failure is synchronous and reserve reactor usage when it is asynchronous.
```

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/VacationResponseGetMethod.scala [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                       mailboxSession: MailboxSession,
+                       processingContext: ProcessingContext): Publisher[Invocation] = {
+    metricFactory.decoratePublisherWithTimerMetricLogP99(JMAP_RFC8621_PREFIX + methodName.value,
+      validate(capabilities)
+        .fold(e => SMono.just(Invocation.error(ErrorCode.UnknownMethod, invocation.methodCallId)),
+          success => asVacationResponseGetRequest(invocation.arguments)
+            .flatMap(vacationResponseGetRequest =>
+              getVacationResponse(vacationResponseGetRequest, processingContext, mailboxSession)
+                .reduce(VacationResponseGetResult(Set.empty, VacationResponseNotFound(Set.empty)), (result1: VacationResponseGetResult, result2: VacationResponseGetResult) => result1.merge(result2))
+                .map(vacationResult => VacationResponseGetResponse(
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/VacationResponseGetMethod.scala:67]: This process method is long to me, maybe we can enhance it via method extraction?

Maybe extracting: `vacationResult => VacationResponseGetResponse` (as a method of vaxationResult???) could help

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/VacationResponseGetMethod.scala [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+                       invocation: Invocation,
+                       mailboxSession: MailboxSession,
+                       processingContext: ProcessingContext): Publisher[Invocation] = {
+    metricFactory.decoratePublisherWithTimerMetricLogP99(JMAP_RFC8621_PREFIX + methodName.value,
+      validate(capabilities)
+        .fold(e => SMono.just(Invocation.error(ErrorCode.UnknownMethod, invocation.methodCallId)),
+          success => asVacationResponseGetRequest(invocation.arguments)
+            .flatMap(vacationResponseGetRequest =>
+              getVacationResponse(vacationResponseGetRequest, processingContext, mailboxSession)
+                .reduce(VacationResponseGetResult(Set.empty, VacationResponseNotFound(Set.empty)), (result1: VacationResponseGetResult, result2: VacationResponseGetResult) => result1.merge(result2))
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/VacationResponseGetMethod.scala:66]: Extract VacationResponseGetResult.empty().

We have no way to shorten `(result1: VacationResponseGetResult, result2: VacationResponseGetResult) => result1.merge(result2)` ?

**Review by chibenwa — APPROVED** [review]

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/VacationResponseGetMethod.scala [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+  private def asVacationResponseGetRequest(arguments: Arguments): Either[IllegalArgumentException, VacationResponseGetRequest] =
+    serializer.deserializeVacationResponseGetRequest(arguments.value) match {
+      case JsSuccess(vacationResponseGetRequest, _) => Right(vacationResponseGetRequest)
+      case errors: JsError => Left(new IllegalArgumentException(serializer.serialize(errors).toString))
+    }
+
+  private def handleRequestValidationErrors(exception: Exception, methodCallId: MethodCallId): Publisher[Invocation] = exception match {
+    case e: MissingCapabilityException => SMono.just(Invocation.error(ErrorCode.UnknownMethod, methodCallId))
```
Comment (remk) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/VacationResponseGetMethod.scala:96]: s/case e:/case _:

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/model/Capability.scala [hunk 10]
```diff
@@ -23,15 +23,19 @@ import eu.timepit.refined.api.Refined
 import eu.timepit.refined.auto._
 import eu.timepit.refined.collection.NonEmpty
 import eu.timepit.refined.string.Uri
-import org.apache.james.jmap.model.CapabilityIdentifier.{CapabilityIdentifier, JMAP_CORE, JMAP_MAIL, JAMES_QUOTA, JAMES_SHARES}
+import org.apache.james.jmap.model.CapabilityIdentifier.{CapabilityIdentifier, JAMES_QUOTA, JAMES_SHARES, JMAP_CORE, JMAP_MAIL, JMAP_VACATION_RESPONSE}
 import org.apache.james.jmap.model.CoreCapabilityProperties.CollationAlgorithm
 import org.apache.james.jmap.model.MailCapability.EmailQuerySortOption
 import org.apache.james.jmap.model.UnsignedInt.UnsignedInt
 
+class CapabilityValidationException() extends IllegalArgumentException
```
Comment (remk) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/model/Capability.scala:31]: could be a sealed trait if not extended elsewhere ?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/model/Capability.scala [hunk 11]
```diff
@@ -86,20 +90,19 @@ final case class MailCapabilityProperties(maxMailboxesPerEmail: MaxMailboxesPerE
...
-final case class SharesCapabilityProperties() extends CapabilityProperties {
-}
+final case class SharesCapabilityProperties() extends CapabilityProperties
 
 final case class SharesCapability(properties: SharesCapabilityProperties = SharesCapabilityProperties(),
-                                  identifier: CapabilityIdentifier = JAMES_SHARES) extends Capability {
-}
+                                  identifier: CapabilityIdentifier = JAMES_SHARES) extends Capability
+
+final case class VacationResponseCapabilityProperties() extends CapabilityProperties
```
Comment (remk) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/model/Capability.scala:105]: case object ?

**Review by remk — APPROVED** [review]

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/VacationResponseGetMethodContract.scala [hunk 12]
```diff
@@ -194,4 +194,137 @@ trait VacationResponseGetMethodContract {
          |    "c1"]]
          |}""".stripMargin)
   }
+
+  @Test
+  def vacationResponseShouldReturnWhenSingletonId(): Unit = {
```
[the remark this answers is not in the document]
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/VacationResponseGetMethodContract.scala:199]: the vacation response? ^^

How about `vacationResponseShouldReturnValidResponseWhenSingletonId`?

File: server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/json/VacationResponseSerializationTest.scala [hunk 13]
```diff
@@ -24,19 +24,20 @@ import java.time.ZonedDateTime
...
 import org.apache.james.mailbox.model.TestId
 import org.scalatest.matchers.should.Matchers
 import org.scalatest.wordspec.AnyWordSpec
 import play.api.libs.json.Json
 
 object VacationResponseSerializationTest {
-  private val DATE_TIME_2016 = ZonedDateTime.parse("2016-10-09T08:07:06+07:00[Asia/Vientiane]")
-  private val DATE_TIME_2017 = ZonedDateTime.parse("2017-10-09T08:07:06+07:00[Asia/Vientiane]")
+  private val UTC_DATE_TIME_2016 = UTCDate(ZonedDateTime.parse("2016-10-09T08:07:06+07:00[Asia/Vientiane]"))
+  private val UTC_DATE_TIME_2017 = UTCDate(ZonedDateTime.parse("2017-10-09T08:07:06+07:00[Asia/Vientiane]"))
```
[the remark this answers is not in the document]
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/json/VacationResponseSerializationTest.scala:35]: To prove we can pass anything we always get it converted to UTC? But if it's confusing I can just use UTC times

[the remark this answers is not in the document]
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/json/VacationResponseSerializationTest.scala:35]: I do prove it in UTCdateTest unit tests though I guess your comment makes sense

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/model/Capability.scala (hunk 11 above)
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/model/Capability.scala:105]: You did similar things with @rouazana regarding `QuotaCapabilityProperties` and `SharesCapabilityProperties`. I'm not expert after but if I remember it had to do with the serialization, as we wanted to have those being sent back as part of the session response even if they are empty?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/VacationResponseGetMethodContract.scala [hunk 14]
```diff
@@ -327,4 +327,85 @@ trait VacationResponseGetMethodContract {
...
+      .stripMargin
+
+    assertThatJson(response).isEqualTo(
+      """{
+         |  "sessionState": "75128aab4b1b",
+         |  "methodResponses": [[
+         |    "error",
+         |      {
+         |        "type": "error",
+         |        "description": "{\"errors\":[{\"path\":\"obj.ids[0]\",\"messages\":[\"Predicate isEmpty() did not fail.\"]}]}"
```
[the remark this answers is not in the document]
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/VacationResponseGetMethodContract.scala:406]: I agree with you it is confusing, but that's a bit the general message taken from the `IllegalArgumentException` when deserializing the VacationResponseGetRequest. There could be other json deserialization errors... Not sure why play json returns such a weird message for empty elements like that and what you want me to do here?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/model/Capability.scala (hunk 11 above)
Comment (remk) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/model/Capability.scala:105]: Ok, keep it like that for the moment. I will see if i can refactor all those empty case classes later.

Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/model/Capability.scala:105]: Yeah sorry I tried to check quickly but couldn't really get to anything

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
Merged
